test/e2e/dr/quorum-restore.go

```diff
@@ -0,0 +1,315 @@
+package dr
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	e2e "k8s.io/kubernetes/test/e2e/framework"
+
+	exutil "github.com/openshift/origin/test/extended/util"
+
+	g "github.com/onsi/ginkgo"
+	o "github.com/onsi/gomega"
+)
+
+const (
+	machineAnnotationName   = "machine.openshift.io/machine"
+	localEtcdSignerYaml     = "/tmp/kube-etcd-cert-signer.yaml"
+	expectedNumberOfMasters = 3
+)
+
+var _ = g.Describe("[Feature:DisasterRecovery][Disruptive]", func() {
+	f := e2e.NewDefaultFramework("disaster-recovery")
+	f.SkipNamespaceCreation = true
+	f.SkipPrivilegedPSPBinding = true
+
+	oc := exutil.NewCLIWithoutNamespace("disaster-recovery")
+
+	g.It("[dr-quorum-restore] Cluster should restore itself after quorum loss", func() {
+		config, err := e2e.LoadConfig()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		dynamicClient := dynamic.NewForConfigOrDie(config)
+		ms := dynamicClient.Resource(schema.GroupVersionResource{
+			Group:    "machine.openshift.io",
+			Version:  "v1beta1",
+			Resource: "machines",
+		}).Namespace("openshift-machine-api")
+
+		bastionHost := setupSSHBastion(oc)
+		proxy := fmt.Sprintf(proxyTemplate, sshOpts, bastionHost)
+		defer removeSSHBastion(oc)
+
+		scaleEtcdQuorum(oc, 0)
+
+		e2e.Logf("Finding two masters to remove")
+		mapiPods, err := oc.AdminKubeClient().CoreV1().Pods("openshift-machine-api").List(metav1.ListOptions{
+			LabelSelector: "k8s-app=controller",
+		})
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(mapiPods.Items).NotTo(o.BeEmpty())
+
+		survivingNodeName := mapiPods.Items[0].Spec.NodeName
```

Comment:
Just a question how random is the list here. Meaning are we making any assumptions by always taking the first in the list. I assume mapiPods is in no predetermined order but I wanted to make sure.

Reply:
The test would pick the first one as there should be just one instance of machine-api-controller pod running

Reply:
OK so we are in a sense not randomizing what survives. I know we need to do this for some reason but the customer does not get to make this choice. Do we need to fix this core problem?

Reply:
If we nuke machine-api-controller the operator should reschedule it to another master.

Reply:
Is this just a matter of waiting for that to happen? can we force it?

Reply:
Correct, it would reschedule to a different master. The issue is that it won't allow to remove a master its running on via Machine API.

So the test would find out which node machine controller is running on and remove the other two - it can't unfortunately kill random masters